In [2]:
2

2

### Finding the Derivatives of Other Functions

- Import the libraries

In [3]:
from math import sqrt

- Defined the equations in Python

In [4]:
def f(x):
    return 6*x**3

def g(x):
    return sqrt(2*x+5)

def h(x):
    return 1/(x-3)**3

- Define the derivative function

In [5]:
def derivative(f,x):
    delta_x = 1/1000000
    return (f(x-delta_x)-f(x))/delta_x

- Print the derivatives

In [6]:
print(derivative(f,-2),derivative(g,3),derivative(h,5))

-72.00003601326443 -0.30151135810285723 0.18750018754287723


- Next define new function

In [7]:
def f(x):
    return x**3 - 2*x**2 + 1

- Write the function to return the intercept of a line given the slope and a point

In [8]:
def point_slope(m,x,y):
    return y-m*x

- Write a function that takes the function f and an x value and finds the derivative of f at x, puts that into the point slope function, and prints out the equation

In [9]:
def tangent_line(f,x):
    m = derivative(f,x)
    y0 = f(x) # Find the intercept
    b = point_slope(m,x,y0)
    print("y = ",round(m,2),"x + ",round(b,2))

- Do this to get the equations

In [10]:
for x in [-0.48,0.67]:
    tangent_line(f,x)

y =  -2.61 x +  -0.82
y =  1.33 x +  -0.49


### Finding the Area Under a Curve

In [13]:
from math import cos,pi

def f(x):
    return x**3 + 3

def g(x):
    return 3*cos(x)

def h(x):
    return ((x**2 - 1)*(x**2+1))/x**2

- Create trap_integral function that return num of trapezoids under f between a and b

In [14]:
def trap_integral(f,a,b,num):
    """Returns the sum of num trapezoids
    under f between a and b"""
    width = (b-a)/num
    area = 0.5*width*(f(a) + f(b) + 2*sum([f(a+width*n) for n in range(1,num)]))
    return area


- Call <b>trap_integral</b> function

In [15]:
print(trap_integral(f,3,4,100))
print(trap_integral(g,0,pi/4,100))
print(trap_integral(h,2,4,100))

46.75017499999999
2.1213094390731206
18.416792708494786


### Finding the Volume of Solid of Revolution

- Create f and g as usual, and third function (h) to be the differene of the squares of f and g

In [16]:
def f(x): return 4-4*x**2
def g(x): return 1-x**2
def h(x): return f(x)**2 - g(x)**2

- Now the volume of solid will be the sum of a given number of cylinders made between the functions. We do this thing as in our integration function. The radius of the cylinder is the same as height of our rectangle when we were integrating

In [17]:
def vol_solid(f,a,b):
    volume = 0
    num = 10000
    width = (b-a)/num
    for i in range(num): # Volume of a cylinder is pi*r^2*h. Run total volume!
        vol = pi*(f(a+i*width))*width
        volume += vol

    return volume

- Call vol_solid on the h funciton for x between -1 and 1. The result of solid volume is 50.3 cubic units

In [19]:
print(vol_solid(h, -1, 1))

50.26548245743666


### Find Quickest Route

Di sini kita belajar menemukan nilai minimum dari satu model fungsi. Biasanya kita harus mengambil turunan secara aljabar, mengset nya ke nilai netral (nol), kemudian menyelesaikan persamaan yang dihasilkan. Ini bisa diautomasi dengan proramming menggunakan model turunan untuk menemukan nilai minimum dengan mudah.

- A lighthouse is located 6 kilometers offshore, and a cabin on the straight shoreline is 9 kilometers from the point on the shore nearest the lighthouse. If you row at a rate of 3 km/hr and walk at a rate of 5 km/hr, where should you land your boat in order to get from the lighthouse to the cabin as quickly as possible?

In [23]:
# There is hte function we need to minimize. The optimal x is going to be between 0 and 9 kilometers
# so we will set those as our start and end values when we call those functions

from math import sqrt

def t(x): return sqrt(x**2 + 36)/3 + (9-x)/5

# Define the function find max mins to start a minimum x value and
# take tiny steps, checking if the derivative equals zero or otherwise
# from positive to negative, or vice versa

def find_max_mins(f,start,stop,step=0.001):
    x = start
    deriv = derivative(f,x)
    while x < stop:
        x += step
        #take derivative at new x:
        newderiv = derivative(f,x)
        #if derivative changes sign
        if newderiv == 0 or deriv*newderiv < 0:
            print("Max/Min at x=",x,"y=",f(x))
            #change deriv to newderiv
            deriv = newderiv

find_max_mins(t,0,9)

Max/Min at x= 4.500999999999838 y= 3.400000014221084


Deket kan? 4.5 kilometer sepanjang pantai. Kalkulasinya di fase ini sangatlah berguna, karean kita juga menemukan path terpendek dari 2 point